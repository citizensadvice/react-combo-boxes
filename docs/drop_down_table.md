# `<DropDownTable>`

An alternative layout for a drop down.  This places the results in a table which is more readable for tabulated data.

This has the same options as [DropDown](drop-down) and [ComboBoxTable](combo-box-table).

## Usage

```js
function map({ name }) {
  return { label: name };
}

const options = [
  { name: 'Frank', age: 21, height: `6'` },
  { name: 'Bob', age: 42, height: `5'11"` },
  { name: 'Harry', age: 1, height: `1'2"` },
];

<DropDownTable
  {...props}
  options={options}
  columns={['name', 'age', 'height']} 
  mapOption={map}
/>
```

[drop-down]: drop_down.md
[combo-box-table]: combo_box_table.md
