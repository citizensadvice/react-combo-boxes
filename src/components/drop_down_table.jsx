import { Fragment, forwardRef, useMemo } from 'react';
import PropTypes from 'prop-types';
import { DropDown } from './drop_down';
import { renderListBox } from './list_box_table/render_list_box';
import { renderGroupLabel } from './list_box_table/render_group_label';
import { renderOption } from './list_box_table/render_option';

function renderNothing() {
  return null;
}

export const DropDownTable = forwardRef(({ columns: rawColumns, ...props }, ref) => {
  const columns = useMemo(() => (
    rawColumns.map((column) => {
      if (typeof column === 'string') {
        return {
          name: column,
        };
      }
      return column;
    })
  ), [rawColumns]);

  return (
    <DropDown
      ref={ref}
      {...props}
      columns={columns}
      renderValue={renderNothing}
      renderListBox={renderListBox}
      renderOption={renderOption}
      renderGroupLabel={renderGroupLabel}
    />
  );
});

DropDownTable.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      label: PropTypes.string,
      html: PropTypes.object,
    }),
  ])).isRequired,
  renderTableWrapper: PropTypes.func,
  renderTable: PropTypes.func,
  renderTableHeaderCell: PropTypes.func,
  renderTableGroupRow: PropTypes.func,
  renderTableGroupHeaderCell: PropTypes.func,
  renderTableRow: PropTypes.func,
  renderTableCellColumnAccessibleLabel: PropTypes.func,
  renderTableCell: PropTypes.func,
  renderColumnValue: PropTypes.func,
};

DropDownTable.defaultProps = {
  renderTableWrapper: (props) => <div {...props} />,
  renderTable: (props) => <table {...props} />,
  renderTableHeaderCell: (props) => <th {...props} />,
  renderTableGroupRow: (props) => <tr {...props} />,
  renderTableGroupHeaderCell: (props) => <th {...props} />,
  renderTableRow: (props) => <tr {...props} />,
  renderTableCellColumnAccessibleLabel: (props) => <span {...props} />,
  renderTableCell: (props) => <td {...props} />,
  // eslint-disable-next-line react/jsx-no-useless-fragment
  renderColumnValue: (props) => <Fragment {...props} />,
};

DropDownTable.displayName = 'DropDownTable';
