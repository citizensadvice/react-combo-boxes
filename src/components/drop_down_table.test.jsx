/* eslint-disable testing-library/no-node-access */

import React, { useState, forwardRef } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DropDownTable } from './drop_down_table';

const DropDownWrapper = forwardRef(({ value: initialValue, ...props }, ref) => {
  const [value, onValue] = useState(initialValue);
  return (
    <DropDownTable
      id="id"
      aria-labelledby="id-label"
      value={value}
      onValue={onValue}
      {...props}
      ref={ref}
    />
  );
});

describe('columns as names only', () => {
  const options = [
    { label: 'Apple', type: 'Fruit', colour: 'Green' },
    { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
    { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
    { label: 'Tomato', type: 'Vegetable', colour: 'Red', disabled: true },
  ];

  const columns = ['label', 'type'];

  it('renders a table as the list box', () => {
    const { container } = render((
      <DropDownWrapper
        options={options}
        columns={columns}
      />
    ));
    expect(container).toMatchSnapshot();
  });

  it('allows selection by click', async () => {
    const spy = jest.fn();
    render((
      <DropDownWrapper
        options={options}
        columns={columns}
        onValue={spy}
      />
    ));
    await userEvent.click(screen.getByRole('combobox'));
    await userEvent.click(screen.getByRole('option', { name: /Banana/ }));
    expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
  });

  it('allows selection by keyboard', async () => {
    const spy = jest.fn();
    render((
      <DropDownWrapper
        options={options}
        columns={columns}
        onValue={spy}
      />
    ));

    screen.getByRole('combobox').focus();
    await userEvent.keyboard('{ArrowDown}{ArrowDown}{Enter}');

    expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
  });
});

describe('columns with headers', () => {
  const options = [
    { label: 'Apple', type: 'Fruit', colour: 'Green' },
    { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
    { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
  ];

  const columns = [
    { name: 'label', label: 'Name' },
    { name: 'type', label: 'Type' },
  ];

  it('renders a table with headers', () => {
    const { container } = render((
      <DropDownWrapper
        options={options}
        columns={columns}
      />
    ));
    expect(container).toMatchSnapshot();
  });

  it('allows selection by click', async () => {
    const spy = jest.fn();
    render((
      <DropDownWrapper
        options={options}
        columns={columns}
        onValue={spy}
      />
    ));
    await userEvent.click(screen.getByRole('combobox'));
    await userEvent.click(screen.getByRole('option', { name: /Banana/ }));
    expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
  });

  it('allows selection by keyboard', async () => {
    const spy = jest.fn();
    render((
      <DropDownWrapper
        options={options}
        columns={columns}
        onValue={spy}
      />
    ));
    screen.getByRole('combobox').focus();
    await userEvent.keyboard('{ArrowDown}{ArrowDown}{Enter}');

    expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
  });
});

describe('columns with html', () => {
  const options = [
    { label: 'Apple', type: 'Fruit', colour: 'Green' },
    { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
    { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
  ];

  const columns = [
    { name: 'label', colHtml: { className: 'foo' } },
    { name: 'type', colHtml: { className: 'bar' } },
  ];

  it('renders a table with colgroups', () => {
    const { container } = render((
      <DropDownWrapper
        options={options}
        columns={columns}
      />
    ));
    expect(container).toMatchSnapshot();
  });
});

describe('cells with html', () => {
  const options = [
    { label: 'Apple', type: 'Fruit', colour: 'Green' },
    { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
    { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
  ];

  const columns = [
    { name: 'label', cellHtml: { className: 'foo' } },
    { name: 'type', cellHtml: { className: 'bar' } },
  ];

  it('renders a table with colgroups', () => {
    const { container } = render((
      <DropDownWrapper
        options={options}
        columns={columns}
      />
    ));
    expect(container).toMatchSnapshot();
  });
});

describe('grouped', () => {
  const options = [
    { name: 'Apple', type: 'Fruit', colour: 'Green' },
    { name: 'Banana', type: 'Fruit', colour: 'Yellow' },
    { name: 'Potato', type: 'Vegetable', colour: 'Brown' },
  ];

  const columns = ['name', 'colour'];

  function map({ name, type }) {
    return {
      label: name,
      group: type,
    };
  }

  it('renders a table with groups', () => {
    const { container } = render((
      <DropDownWrapper
        options={options}
        columns={columns}
        mapOption={map}
      />
    ));
    expect(container).toMatchSnapshot();
  });

  it('allows selection by click', async () => {
    const spy = jest.fn();
    render((
      <DropDownWrapper
        options={options}
        columns={columns}
        onValue={spy}
        mapOption={map}
      />
    ));
    await userEvent.click(screen.getByRole('combobox'));
    await userEvent.click(screen.getByRole('option', { name: /Banana/ }));
    expect(spy).toHaveBeenCalledWith({ name: 'Banana', type: 'Fruit', colour: 'Yellow' });
  });

  it('allows selection by keyboard', async () => {
    const spy = jest.fn();
    render((
      <DropDownWrapper
        options={options}
        columns={columns}
        onValue={spy}
        mapOption={map}
      />
    ));
    screen.getByRole('combobox').focus();
    await userEvent.keyboard('{ArrowDown}{ArrowDown}{Enter}');

    expect(spy).toHaveBeenCalledWith({ name: 'Banana', type: 'Fruit', colour: 'Yellow' });
  });
});
