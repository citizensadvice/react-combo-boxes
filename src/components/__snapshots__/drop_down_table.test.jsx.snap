// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cells with html renders a table with colgroups 1`] = `
<div>
  <div
    class="react-combo-boxes-dropdown"
  >
    <div
      aria-controls="id_listbox"
      aria-describedby="id_aria_description"
      aria-expanded="false"
      aria-labelledby="id-label id"
      class="react-combo-boxes-dropdown__combobox"
      id="id"
      role="combobox"
      tabindex="0"
    >
      Apple
    </div>
    <div
      class="react-combo-boxes-dropdown__listbox-wrapper"
    >
      <div
        class="react-combo-boxes-dropdown__listbox react-combo-boxes-dropdown__false"
        hidden=""
        tabindex="-1"
      >
        <table
          aria-labelledby="id-label"
          class="react-combo-boxes-dropdown__table"
          id="id_listbox"
          role="listbox"
          tabindex="-1"
        >
          <colgroup>
            <col />
            <col />
          </colgroup>
          <tbody
            role="presentation"
          >
            <tr
              aria-labelledby="id_option_apple_cell_label id_option_apple_cell_type"
              aria-selected="true"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_apple"
              role="option"
              tabindex="-1"
            >
              <td
                class="foo"
                id="id_option_apple_cell_label"
                role="presentation"
              >
                Apple
              </td>
              <td
                class="bar"
                id="id_option_apple_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_option_banana_cell_label id_option_banana_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_banana"
              role="option"
              tabindex="-1"
            >
              <td
                class="foo"
                id="id_option_banana_cell_label"
                role="presentation"
              >
                Banana
              </td>
              <td
                class="bar"
                id="id_option_banana_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_option_potato_cell_label id_option_potato_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_potato"
              role="option"
              tabindex="-1"
            >
              <td
                class="foo"
                id="id_option_potato_cell_label"
                role="presentation"
              >
                Potato
              </td>
              <td
                class="bar"
                id="id_option_potato_cell_type"
                role="presentation"
              >
                Vegetable
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_aria_description"
    >
      When results are available use up and down arrows to review and enter to select
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        role="status"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`columns as names only renders a table as the list box 1`] = `
<div>
  <div
    class="react-combo-boxes-dropdown"
  >
    <div
      aria-controls="id_listbox"
      aria-describedby="id_aria_description"
      aria-expanded="false"
      aria-labelledby="id-label id"
      class="react-combo-boxes-dropdown__combobox"
      id="id"
      role="combobox"
      tabindex="0"
    >
      Apple
    </div>
    <div
      class="react-combo-boxes-dropdown__listbox-wrapper"
    >
      <div
        class="react-combo-boxes-dropdown__listbox react-combo-boxes-dropdown__false"
        hidden=""
        tabindex="-1"
      >
        <table
          aria-labelledby="id-label"
          class="react-combo-boxes-dropdown__table"
          id="id_listbox"
          role="listbox"
          tabindex="-1"
        >
          <colgroup>
            <col />
            <col />
          </colgroup>
          <tbody
            role="presentation"
          >
            <tr
              aria-labelledby="id_option_apple_cell_label id_option_apple_cell_type"
              aria-selected="true"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_apple"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_label"
                role="presentation"
              >
                Apple
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_option_banana_cell_label id_option_banana_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_banana"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_label"
                role="presentation"
              >
                Banana
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_option_potato_cell_label id_option_potato_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_potato"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_label"
                role="presentation"
              >
                Potato
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_type"
                role="presentation"
              >
                Vegetable
              </td>
            </tr>
            <tr
              aria-disabled="true"
              aria-labelledby="id_option_tomato_cell_label id_option_tomato_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_tomato"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_tomato_cell_label"
                role="presentation"
              >
                Tomato
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_tomato_cell_type"
                role="presentation"
              >
                Vegetable
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_aria_description"
    >
      When results are available use up and down arrows to review and enter to select
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        role="status"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`columns with headers and grouped renders a table with headers 1`] = `
<div>
  <div
    class="react-combo-boxes-dropdown"
  >
    <div
      aria-controls="id_listbox"
      aria-describedby="id_aria_description"
      aria-expanded="false"
      aria-labelledby="id-label id"
      class="react-combo-boxes-dropdown__combobox"
      id="id"
      role="combobox"
      tabindex="0"
    >
      Â 
    </div>
    <div
      class="react-combo-boxes-dropdown__listbox-wrapper"
    >
      <div
        class="react-combo-boxes-dropdown__listbox react-combo-boxes-dropdown__listbox--header"
        hidden=""
        tabindex="-1"
      >
        <table
          aria-labelledby="id-label"
          class="react-combo-boxes-dropdown__table"
          id="id_listbox"
          role="listbox"
          tabindex="-1"
        >
          <colgroup>
            <col />
            <col />
          </colgroup>
          <thead
            role="presentation"
          >
            <tr
              role="presentation"
            >
              <th
                class="react-combo-boxes-dropdown__table-header"
                id="id_column_label"
              >
                Name
              </th>
              <th
                class="react-combo-boxes-dropdown__table-header"
                id="id_column_colour"
              >
                Colour
              </th>
            </tr>
          </thead>
          <tbody
            role="presentation"
          >
            <tr
              class="react-combo-boxes-dropdown__table-group-row"
            >
              <th
                aria-hidden="true"
                class="react-combo-boxes-dropdown__table-group-header"
                colspan="2"
                id="id_group_fruit"
              >
                Fruit
              </th>
            </tr>
            <tr
              aria-labelledby="id_group_fruit id_column_label id_option__cell_label id_column_colour id_option__cell_colour"
              aria-selected="true"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option__cell_label"
                role="presentation"
              >
                Apple
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option__cell_colour"
                role="presentation"
              >
                Green
              </td>
            </tr>
            <tr
              aria-labelledby="id_group_fruit id_column_label id_option_1_cell_label id_column_colour id_option_1_cell_colour"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_1"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_1_cell_label"
                role="presentation"
              >
                Banana
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_1_cell_colour"
                role="presentation"
              >
                Yellow
              </td>
            </tr>
            <tr
              class="react-combo-boxes-dropdown__table-group-row"
            >
              <th
                aria-hidden="true"
                class="react-combo-boxes-dropdown__table-group-header"
                colspan="2"
                id="id_group_vegetable"
              >
                Vegetable
              </th>
            </tr>
            <tr
              aria-labelledby="id_group_vegetable id_column_label id_option_2_cell_label id_column_colour id_option_2_cell_colour"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_2"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_2_cell_label"
                role="presentation"
              >
                Potato
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_2_cell_colour"
                role="presentation"
              >
                Brown
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_aria_description"
    >
      When results are available use up and down arrows to review and enter to select
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        role="status"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`columns with headers renders a table with headers 1`] = `
<div>
  <div
    class="react-combo-boxes-dropdown"
  >
    <div
      aria-controls="id_listbox"
      aria-describedby="id_aria_description"
      aria-expanded="false"
      aria-labelledby="id-label id"
      class="react-combo-boxes-dropdown__combobox"
      id="id"
      role="combobox"
      tabindex="0"
    >
      Apple
    </div>
    <div
      class="react-combo-boxes-dropdown__listbox-wrapper"
    >
      <div
        class="react-combo-boxes-dropdown__listbox react-combo-boxes-dropdown__listbox--header"
        hidden=""
        tabindex="-1"
      >
        <table
          aria-labelledby="id-label"
          class="react-combo-boxes-dropdown__table"
          id="id_listbox"
          role="listbox"
          tabindex="-1"
        >
          <colgroup>
            <col />
            <col />
          </colgroup>
          <thead
            role="presentation"
          >
            <tr
              role="presentation"
            >
              <th
                class="react-combo-boxes-dropdown__table-header"
                id="id_column_label"
              >
                Name
              </th>
              <th
                class="react-combo-boxes-dropdown__table-header"
                id="id_column_type"
              >
                Type
              </th>
            </tr>
          </thead>
          <tbody
            role="presentation"
          >
            <tr
              aria-labelledby="id_column_label id_option_apple_cell_label id_column_type id_option_apple_cell_type"
              aria-selected="true"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_apple"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_label"
                role="presentation"
              >
                Apple
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_column_label id_option_banana_cell_label id_column_type id_option_banana_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_banana"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_label"
                role="presentation"
              >
                Banana
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_column_label id_option_potato_cell_label id_column_type id_option_potato_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_potato"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_label"
                role="presentation"
              >
                Potato
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_type"
                role="presentation"
              >
                Vegetable
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_aria_description"
    >
      When results are available use up and down arrows to review and enter to select
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        role="status"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`columns with html renders a table with colgroups 1`] = `
<div>
  <div
    class="react-combo-boxes-dropdown"
  >
    <div
      aria-controls="id_listbox"
      aria-describedby="id_aria_description"
      aria-expanded="false"
      aria-labelledby="id-label id"
      class="react-combo-boxes-dropdown__combobox"
      id="id"
      role="combobox"
      tabindex="0"
    >
      Apple
    </div>
    <div
      class="react-combo-boxes-dropdown__listbox-wrapper"
    >
      <div
        class="react-combo-boxes-dropdown__listbox react-combo-boxes-dropdown__false"
        hidden=""
        tabindex="-1"
      >
        <table
          aria-labelledby="id-label"
          class="react-combo-boxes-dropdown__table"
          id="id_listbox"
          role="listbox"
          tabindex="-1"
        >
          <colgroup>
            <col
              class="foo"
            />
            <col
              class="bar"
            />
          </colgroup>
          <tbody
            role="presentation"
          >
            <tr
              aria-labelledby="id_option_apple_cell_label id_option_apple_cell_type"
              aria-selected="true"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_apple"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_label"
                role="presentation"
              >
                Apple
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_option_banana_cell_label id_option_banana_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_banana"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_label"
                role="presentation"
              >
                Banana
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_type"
                role="presentation"
              >
                Fruit
              </td>
            </tr>
            <tr
              aria-labelledby="id_option_potato_cell_label id_option_potato_cell_type"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_potato"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_label"
                role="presentation"
              >
                Potato
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_type"
                role="presentation"
              >
                Vegetable
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_aria_description"
    >
      When results are available use up and down arrows to review and enter to select
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        role="status"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`grouped renders a table with groups 1`] = `
<div>
  <div
    class="react-combo-boxes-dropdown"
  >
    <div
      aria-controls="id_listbox"
      aria-describedby="id_aria_description"
      aria-expanded="false"
      aria-labelledby="id-label id"
      class="react-combo-boxes-dropdown__combobox"
      id="id"
      role="combobox"
      tabindex="0"
    >
      Apple
    </div>
    <div
      class="react-combo-boxes-dropdown__listbox-wrapper"
    >
      <div
        class="react-combo-boxes-dropdown__listbox react-combo-boxes-dropdown__false"
        hidden=""
        tabindex="-1"
      >
        <table
          aria-labelledby="id-label"
          class="react-combo-boxes-dropdown__table"
          id="id_listbox"
          role="listbox"
          tabindex="-1"
        >
          <colgroup>
            <col />
            <col />
          </colgroup>
          <tbody
            role="presentation"
          >
            <tr
              class="react-combo-boxes-dropdown__table-group-row"
            >
              <th
                aria-hidden="true"
                class="react-combo-boxes-dropdown__table-group-header"
                colspan="2"
                id="id_group_fruit"
              >
                Fruit
              </th>
            </tr>
            <tr
              aria-labelledby="id_group_fruit id_option_apple_cell_name id_option_apple_cell_colour"
              aria-selected="true"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_apple"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_name"
                role="presentation"
              >
                Apple
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_apple_cell_colour"
                role="presentation"
              >
                Green
              </td>
            </tr>
            <tr
              aria-labelledby="id_group_fruit id_option_banana_cell_name id_option_banana_cell_colour"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_banana"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_name"
                role="presentation"
              >
                Banana
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_banana_cell_colour"
                role="presentation"
              >
                Yellow
              </td>
            </tr>
            <tr
              class="react-combo-boxes-dropdown__table-group-row"
            >
              <th
                aria-hidden="true"
                class="react-combo-boxes-dropdown__table-group-header"
                colspan="2"
                id="id_group_vegetable"
              >
                Vegetable
              </th>
            </tr>
            <tr
              aria-labelledby="id_group_vegetable id_option_potato_cell_name id_option_potato_cell_colour"
              class="react-combo-boxes-dropdown__table-row"
              id="id_option_potato"
              role="option"
              tabindex="-1"
            >
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_name"
                role="presentation"
              >
                Potato
              </td>
              <td
                class="react-combo-boxes-dropdown__table-cell"
                id="id_option_potato_cell_colour"
                role="presentation"
              >
                Brown
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_aria_description"
    >
      When results are available use up and down arrows to review and enter to select
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        role="status"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;
